pkg/gadgets/trace/byob/tracer/tracer.go

```diff
@@ -0,0 +1,346 @@
+		t.ringbufReader.Close()
+	}
+	if t.perfReader != nil {
+		t.perfReader.Close()
+	}
+}
+
+func (t *Tracer) installTracer() error {
+	// Load the spec
+	progReader := bytes.NewReader(t.config.ProgContent)
+	var err error
+	t.spec, err = ebpf.LoadCollectionSpecFromReader(progReader)
+	if err != nil {
+		return fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	mapReplacements := map[string]*ebpf.Map{}
+
+	// Find the print map
+	for mapName, m := range t.spec.Maps {
+		// TODO: Print maps only with prefix print_ ?
+		if (m.Type == ebpf.RingBuf || m.Type == ebpf.PerfEventArray) && strings.HasPrefix(m.Name, "print_") {
+			if t.printMap != "" {
+				return fmt.Errorf("multiple print maps: %q and %q", t.printMap, mapName)
+			}
+			t.printMap = mapName
+
+			var ok bool
+			t.valueStruct, ok = m.Value.(*btf.Struct)
+			if !ok {
+				return fmt.Errorf("BPF map %q does not have BTF info for values", mapName)
+			}
+
+			// Almost same hack as in bumblebee/pkg/loader/loader.go
+			t.mapSizes[mapName] = t.spec.Maps[mapName].ValueSize
+			if m.Type == ebpf.RingBuf {
+				t.spec.Maps[mapName].ValueSize = 0
+			} else if m.Type == ebpf.PerfEventArray {
+				t.spec.Maps[mapName].KeySize = 4
+				t.spec.Maps[mapName].ValueSize = 4
+			}
+		}
+	}
+	if t.printMap == "" {
+		return fmt.Errorf("no BPF map with 'print_' prefix found")
+	}
+
+	// Load the ebpf objects
+	opts := ebpf.CollectionOptions{
+		Programs: ebpf.ProgramOptions{
+			LogSize: ebpf.DefaultVerifierLogSize * 5000,
+		},
+		MapReplacements: mapReplacements,
+	}
+	t.collection, err = ebpf.NewCollectionWithOptions(t.spec, opts)
+	if err != nil {
+		var errVerifier *ebpf.VerifierError
+		if errors.As(err, &errVerifier) {
+			fmt.Printf("Verifier error: %+v\n",
+				errVerifier)
+		}
+		return fmt.Errorf("failed to create BPF collection: %w", err)
+	}
+
+	m := t.collection.Maps[t.printMap]
+	if m.Type() == ebpf.RingBuf {
+		t.ringbufReader, err = ringbuf.NewReader(t.collection.Maps[t.printMap])
+	} else if m.Type() == ebpf.PerfEventArray {
+		t.perfReader, err = perf.NewReader(t.collection.Maps[t.printMap], gadgets.PerfBufferPages*os.Getpagesize())
+	} else {
+		return fmt.Errorf("unsupported BPF map type: %s", m.Type())
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create BPF map reader: %w", err)
+	}
+	go t.run()
+
+	// Attach programs
+	for progName, p := range t.spec.Programs {
+		if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kprobe/") {
+			l, err := link.Kprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("failed to attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		} else if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kretprobe/") {
+			l, err := link.Kretprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("failed to attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		}
+	}
```

Comment:
This is interesting. It's like the autoattach feature of libbpf. I think we could use something similar for other gadgets to avoid attaching all the programs manually.